R/summary.R

```diff
@@ -321,431 +284,139 @@ summary.gs_design <- function(object,
+  # Set the decimals to display ----
+  for (j in col_vars) output[[j]] <- round2(output[[j]], col_decimals[j])
 
+  output <- add_class(
+    output, method, intersect("non_binding", class(object)), method, "gs_design"
+  )
 
-  output <- output %>% select(x_decimals$col_vars)
-  if ("Z" %in% colnames(output)) {
-    output <- output %>% dplyr::mutate_at(
-      "Z",
-      round,
-      (x_decimals %>% filter(col_vars == "Z"))$col_decimals
-    )
-  }
-  if ("~HR at bound" %in% colnames(output)) {
-    output <- output %>% dplyr::mutate_at(
-      "~HR at bound",
-      round,
-      (x_decimals %>% filter(col_vars == "~HR at bound"))$col_decimals
-    )
-  }
-  if ("~Risk difference at bound" %in% colnames(output)) {
-    output <- output %>% dplyr::mutate_at(
-      "~Risk difference at bound",
-      round,
-      (x_decimals %>% filter(col_vars == "~Risk difference at bound"))$col_decimals
-    )
-  }
-  if ("Nominal p" %in% colnames(output)) {
-    output <- output %>% dplyr::mutate_at(
-      "Nominal p",
-      round,
-      (x_decimals %>% filter(col_vars == "Nominal p"))$col_decimals
-    )
-  }
-  if ("Alternate hypothesis" %in% colnames(output)) {
-    output <- output %>% dplyr::mutate_at(
-      "Alternate hypothesis",
-      round,
-      (x_decimals %>% filter(col_vars == "Alternate hypothesis"))$col_decimals
-    )
-  }
-  if ("Null hypothesis" %in% colnames(output) && is.vector(output[["Null hypothesis"]], mode = "numeric")) {
-    output <- output %>% dplyr::mutate_at(
-      "Null hypothesis",
-      round,
-      (x_decimals %>% filter(col_vars == "Null hypothesis"))$col_decimals
-    )
+  # Save the full alpha as an attribute of the output summary table
+  # Use input$alpha when given power to calculate sample size
+  attr(output, "full_alpha") <- object$input$alpha %||% {
+    # when given sample size to calculate power
+    if (is.list(object$input$upar)) object$input$upar$total_spend else 0.025
   }
 
-  class(output) <- c(method, "gs_design", class(output))
-  if ("non_binding" %in% class(object)) {
-    class(output) <- c("non_binding", class(output))
-  }
+  return(output)
+}
 
-  # Save the full alpha as an attribute of the output summary table
-  attr(output, "full_alpha") <-
-    if (is.null(object$input$alpha)) {
-      # case when given sample size to calculate power
-      if (!is.list(object$input$upar)) {
-        0.025
-      } else {
-        object$input$upar$total_spend
-      }
-    } else {
-      # case when given power to calcuate sample size
-      object$input$alpha
-    }
+# get a named vector of decimals (names are variable names)
+get_decimals <- function(vars, decs, vars_default, decs_default) {
+  names(decs_default) <- vars_default
+  # Merge user-provided named decimals into default
+  decs_vars <- names(decs)
+  decs_default[decs_vars] <- decs
+
+  vars_name <- as.character(substitute(vars))
+  decs_name <- as.character(substitute(decs))
```

Comment:
Could you please explain why it is necessary to run `substitute()` here? It isn't immediately obvious to me

Reply:
Sure. I have added some `#` comments. Basically it was a trick to get the argument names `analysis_vars`, `analysis_decimals`, `col_vars`, and `col_decimals`, which will be used in the error messages below. Without the original argument names, the error messages will be less clear. We want an error message like this:

```
'col_decimals' must be a named vector if 'col_vars' is not provided
```

instead of

```
'decs' must be a named vector if 'vars' is not provided
```

Reply:
Ah, thanks for the explanation. I see now that this was purely for the error message, which in retrospect makes total sense.